File: tools/usage-poc/coconut_collector/gitcount.py
"""gitcount.py — device-local git commit count for the VES numerator.

Counts the operator's OWN commits across the repos they worked in during the
reporting window, returning a single integer (the VES numerator) — or None
when the count cannot be trusted.

PRIVACY (absolute): cwd, repo root, commit SHA, and author email are used
ONLY locally to compute one integer. None of them are ever emitted into the
Burn Summary envelope — build_envelope() copies only whitelisted fields and
attaches at most the bare integer `verifiedCommits`.

UNKNOWN != ZERO: when NO working directory resolves to a usable git repo, or
a confirmed git repo errors out mid-count, or there is no author identity, the
result is None ("unknown" → the collector omits the field → UI "—"). A real 0
means at least one repo was inspected cleanly and had zero matching commits.

PARTIAL ATTRIBUTION (conservative undercount): a cwd that is NOT inside a git
work tree (a non-repo wrapper dir, ~, a removed worktree) or is a shallow clone
is SKIPPED rather than poisoning the whole count — its commits are simply
unattributable, so the numerator counts only what the verifiable repos prove.
This makes the VES numerator a conservative lower bound (denominator/cost still
covers all work) — understating, never inflating, and non-gameable. A genuine
git error on a repo that DID resolve still poisons to None (see count_commits).

KNOWN LIMITATION (v1): commits are counted from HEAD only. If the operator made
in-window commits on a feature branch and checked out another branch before
running the collector, those commits are not reachable from HEAD and go
uncounted — so the numerator is checkout-dependent. Widening the ref scope to
local branches (`--branches`) is deferred to v2 (plan Q6, numerator
refinements), since it interacts with the unborn-HEAD probe (orphan branches)
and changes scoring semantics that the S3.5 design ratified as HEAD-only.
"""

import subprocess
from datetime import datetime, timezone

_GIT_TIMEOUT = 30  # seconds; a hung git invocation must not stall the CLI run


def _run_git(args: list[str]) -> "subprocess.CompletedProcess | None":
    """Run a git command, returning the completed process or None on failure
    to even launch (git missing, OS error, timeout, or an un-launchable arg
    such as a cwd carrying the _UNKNOWN_CWD NUL sentinel — ValueError). A cwd
    git cannot even be invoked on is 'unknown' → the caller skips it."""
    try:
        return subprocess.run(
            ["git", *args],
            capture_output=True, text=True, timeout=_GIT_TIMEOUT,
        )
    except (OSError, ValueError, subprocess.SubprocessError):
        return None


def git_author_email() -> str | None:
    """The configured commit author email (`git config user.email`), or None.

    Used ONLY to filter the operator's own commits. Never uploaded.
    """
    proc = _run_git(["config", "user.email"])
    if proc is None or proc.returncode != 0:
        return None
    email = proc.stdout.strip()
    return email or None


def _repo_root(cwd: str) -> str | None:
    """Resolve the git top-level for `cwd`, or None when `cwd` is not inside a
    usable git work tree."""
    proc = _run_git(["-C", cwd, "rev-parse", "--show-toplevel"])
    if proc is None or proc.returncode != 0:
        return None
    root = proc.stdout.strip()
    return root or None


def _is_shallow(root: str) -> "bool | None":
    """Tri-state shallow probe of an ALREADY-RESOLVED `root`.

    True  → shallow clone (truncated history) → caller SKIPS this repo
            (its commit count cannot be trusted → conservative undercount).
    False → full clone (history is trustworthy) → caller counts it.
    None  → git errored on a repo that already resolved at _repo_root()
            → caller POISONS (unknown), per the resolved-repo-error contract.

    The error case must NOT collapse to True (skip): once _repo_root() has
    proven this is a usable work tree, any later git failure on it is a genuine
    error, and a genuine error on a resolved repo is "unknown", not a benign
    skip — otherwise an errored repo would silently drop out of the count while
    another valid repo emits a misleading partial number (codex [P1] 2026-05-29).
    """
    proc = _run_git(["-C", root, "rev-parse", "--is-shallow-repository"])
    if proc is None or proc.returncode != 0:
        return None
    return proc.stdout.strip() == "true"


def _head_state(root: str) -> "bool | None":
    """Tri-state probe of `root`'s HEAD.

    True  → HEAD resolves to at least one commit (proceed to count).
    False → unborn HEAD: a freshly-init'd repo, inspected, zero commits (real 0).
    None  → git could not answer (timeout, OS error, or an unexpected failure
            after the work tree was already resolved) → unknown, omit the field.

    `git rev-parse --verify --quiet HEAD` exits 0 with a born HEAD and 1 for an
    unborn HEAD; any other exit code (e.g. 128 from a corrupt/unreadable repo)
    or a failure to launch git is treated as unknown rather than a real 0 — a
    git error must not masquerade as "the operator made no commits."
    """
    proc = _run_git(["-C", root, "rev-parse", "--verify", "--quiet", "HEAD"])
    if proc is None:
        return None
    if proc.returncode == 0:
        return True
    if proc.returncode == 1:
        return False
    return None


def _parse_iso(value: str) -> "datetime | None":
    """Parse a git %aI author date (strict ISO-8601) into an aware UTC datetime."""
    try:
        dt = datetime.fromisoformat(value.strip())
    except ValueError:
        return None
    if dt.tzinfo is None:
        return dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def _repo_shas(root: str, author_email: str,
               since: datetime, until: datetime) -> "set[str] | None":
    """Distinct SHAs in `root` authored by `author_email` within the half-open
    window [since, until). None on git error.

    The window is enforced IN CODE on the parsed %aI author date rather than via
    `git log --since/--until`, whose fuzzy date parsing can drift by a second at
    the boundary. `--no-use-mailmap` keeps %ae the raw configured email so the
    exact-string author match is deterministic regardless of repo .mailmap.
    """
    head = _head_state(root)
    if head is None:
        return None   # git error after resolving the root → unknown
    if not head:
        return set()  # unborn HEAD: inspected, zero commits — a real 0
    proc = _run_git([
        "-C", root, "log", "--no-use-mailmap",
        "--pretty=format:%H%x09%ae%x09%aI", "HEAD",
    ])
    if proc is None or proc.returncode != 0:
        return None
    shas: set[str] = set()
    for line in proc.stdout.splitlines():
        if not line:
            continue
        parts = line.split("\t")
        if len(parts) != 3:
            continue
        sha, email, authored = parts
        if email != author_email:
            continue
        dt = _parse_iso(authored)
        if dt is None:
            continue
        if since <= dt < until:
            shas.add(sha)
    return shas


def count_commits(repo_cwds, author_email: "str | None",
                  since_utc: datetime, until_utc: datetime) -> "int | None":
    """Distinct commits authored by `author_email` in [since_utc, until_utc)
    across the git repos containing `repo_cwds`.

    repo_cwds: iterable of working directories from the in-window sessions.
      Empty strings, non-git dirs, and shallow clones are SKIPPED (their commits
      are unattributable → conservative undercount, never a poison). Distinct
      repos are counted once even when many sessions share a repo; SHAs are
      de-duplicated across repos before summing.

    Returns None ("unknown", → omit the field) when:
      - author_email is missing (no git identity), or
      - NO supplied cwd resolves to a usable git repo (nothing verifiable), or
      - a repo that DID resolve errors out while counting (genuine git error).
    Returns an int (possibly 0) when at least one repo was inspected cleanly —
    a real 0 means the verifiable repos had no matching commits in the window.
    """
    if not author_email:
        return None
    roots: set[str] = set()
    for cwd in repo_cwds:
        if not cwd:
            continue  # missing cwd → unattributable, skip (not a poison)
        root = _repo_root(cwd)
        if root is None:
            continue  # not a git work tree → unattributable, skip
        shallow = _is_shallow(root)
        if shallow is None:
            return None   # resolved repo errored on shallow probe → poison
        if shallow:
            continue  # truncated history → cannot trust THIS repo, skip it
        roots.add(root)
    if not roots:
        return None  # nothing verifiable → unknown, not zero
    all_shas: set[str] = set()
    for root in roots:
        shas = _repo_shas(root, author_email, since_utc, until_utc)
        if shas is None:
            return None
        all_shas |= shas
    return len(all_shas)

File: lib/client/burn/gitcount.ts
// lib/client/burn/gitcount.ts
//
// Device-local git commit count for the VES numerator — the BROWSER analog of
// tools/usage-poc/coconut_collector/gitcount.py. Counts the operator's OWN
// commits (author email ∈ a selected set) reachable from HEAD across the
// granted repos, within the half-open reporting window [since, until),
// returning a single integer (the VES numerator) — or null when the count
// cannot be trusted.
//
// SEMANTICS are a faithful port of gitcount.py (the authoritative spec):
//   • HEAD ancestry ONLY. A commit made in-window on a feature branch the
//     operator later checked out of is not reachable from HEAD → uncounted.
//     v1 limitation by design (matches the Python collector).
//   • Author email EXACT match, NO mailmap (isomorphic-git applies none, which
//     matches the Python `git log --no-use-mailmap`).
//   • Window enforced IN CODE on each commit's author instant (epoch seconds →
//     ms), never via a fuzzy `git --since`, so the boundary is exact:
//     since <= t < until (half-open).
//   • SHAs de-duplicated GLOBALLY across repos: a commit is content-addressed,
//     so the same SHA seen in two granted repos is one commit (matches the
//     Python `all_shas |= shas`).
//   • UNKNOWN (null) != ZERO. null when: no author identity, OR no granted dir
//     resolves to a usable repo, OR a RESOLVED repo errors mid-count (poison).
//     A real 0 means at least one repo was inspected cleanly with no matching
//     commits (including an unborn HEAD — init'd, no commits).
//   • CONSERVATIVE UNDERCOUNT: a non-git / missing / shallow dir is SKIPPED
//     (its commits are unattributable), never a poison — the numerator is a
//     lower bound, never inflated, and non-gameable. Only a RESOLVED repo that
//     then errors poisons the whole count to null.
//
// PRIVACY (absolute): every input here — the granted directory path strings,
// the author emails, the commit SHAs — is used ONLY to compute one integer.
// NONE of them is ever emitted into the Burn Summary envelope, the POST body,
// or any log. The caller attaches at most the bare integer `verifiedCommits`
// plus the source tag "browser-fsa". See JoinBurnIndexForm.
//
// PERFORMANCE: repos are walked sequentially with an optional progress/abort
// hook. The full HEAD history is scanned per repo and filtered in code — there
// is NO early termination, because author dates are not monotonic along commit
// ancestry, so stopping early on a stale author date could silently undercount.
// Ref-SHA caching and windowed traversal are deferred (v2): correctness over
// speed for v1.
//
// BUNDLE: isomorphic-git is imported lazily (await import) the first time a
// count runs, so it never lands in the default client bundle — only a user who
// clicks "count commits" pays for it.

// --- isomorphic-git surface we use (structurally typed) --------------------
// Only the read-only functions are needed. Kept minimal so the lazy import is
// the single source of the real types at runtime.
interface GitAuthor {
  name: string;
  email: string;
  timestamp: number; // epoch SECONDS, UTC (timezoneOffset is display-only)
  timezoneOffset: number;
}
interface ReadCommitResult {
  oid: string;
  commit: { author: GitAuthor; committer: GitAuthor; message: string };
}
interface GitApi {
  findRoot(args: { fs: GitFs; filepath: string }): Promise<string>;
  resolveRef(args: { fs: GitFs; dir: string; ref: string }): Promise<string>;
  log(args: { fs: GitFs; dir: string; ref: string }): Promise<ReadCommitResult[]>;
  getConfig(args: { fs: GitFs; dir: string; path: string }): Promise<string | undefined>;
  Errors: { NotFoundError: new (...a: never[]) => Error };
}

// --- fs the adapter must present (validated by spike, 2026-05-30) ----------
// bindFs() binds the full command list at construction, so EVERY method name
// must EXIST as a function — but only readFile + stat are CALLED on the
// read-only log/findRoot/resolveRef path. Write ops are present-but-throwing.
export interface GitStats {
  type: "file" | "dir";
  mode: number;
  size: number;
  ino: number;
  dev: number;
  uid: number;
  gid: number;
  ctimeMs: number;
  mtimeMs: number;
  ctimeSeconds: number;
  ctimeNanoseconds: number;
  mtimeSeconds: number;
  mtimeNanoseconds: number;
  isFile(): boolean;
  isDirectory(): boolean;
  isSymbolicLink(): boolean;
}
export interface GitFsPromises {
  readFile(path: string, options?: { encoding?: string } | string): Promise<Uint8Array | string>;
  stat(path: string): Promise<GitStats>;
  lstat(path: string): Promise<GitStats>;
  readdir(path: string): Promise<string[]>;
  readlink(path: string): Promise<string>;
  writeFile(...args: unknown[]): Promise<void>;
  mkdir(...args: unknown[]): Promise<void>;
  unlink(...args: unknown[]): Promise<void>;
  symlink(...args: unknown[]): Promise<void>;
  rmdir(path: string, options?: unknown): Promise<void>;
}
export interface GitFs {
  promises: GitFsPromises;
}

// --- lazy isomorphic-git loader --------------------------------------------
let gitModulePromise: Promise<GitApi> | null = null;

/** Lazily import isomorphic-git. The ESM build exports `default` (an object
 *  carrying all functions) AND named exports; the CJS build (Node/vitest) wraps
 *  module.exports under `default`. Pick whichever namespace actually has
 *  `.log`. Cache the resolved module; reset on failure so a transient import
 *  error does not poison every later call. */
export function loadGit(): Promise<GitApi> {
  if (gitModulePromise == null) {
    gitModulePromise = import("isomorphic-git")
      .then((mod) => {
        const ns = mod as unknown as Record<string, unknown> & { default?: Record<string, unknown> };
        const picked = ns.default && typeof ns.default.log === "function" ? ns.default : ns;
        return picked as unknown as GitApi;
      })
      .catch((err) => {
        gitModulePromise = null;
        throw err;
      });
  }
  return gitModulePromise;
}

// --- tiny POSIX path join (isomorphic-git builds paths with "/") -----------
function joinPath(...parts: string[]): string {
  return parts
    .map((p, i) => (i === 0 ? p.replace(/\/+$/, "") : p.replace(/^\/+|\/+$/g, "")))
    .filter((p) => p.length > 0)
    .join("/");
}

function isENOENT(err: unknown): boolean {
  const code = (err as { code?: string } | null)?.code;
  return code === "ENOENT" || code === "ENOTDIR";
}

// --- per-repo probes (faithful ports of the Python tri-states) -------------

/** Resolve `dir` to its repo root (analog of `git rev-parse --show-toplevel`),
 *  or null when `dir` is not inside a usable git work tree. ANY resolution
 *  failure (NotFoundError walking up, or an fs read error) is a SKIP, not a
 *  poison — mirrors Python `_repo_root` returning None → caller skips. */
async function findRepoRoot(git: GitApi, fs: GitFs, dir: string): Promise<string | null> {
  try {
    return await git.findRoot({ fs, filepath: dir });
  } catch {
    return null;
  }
}

/** Tri-state shallow probe of an ALREADY-RESOLVED root, by the existence of
 *  `<root>/.git/shallow`:
 *    true  → shallow clone (truncated history) → caller SKIPS this repo.
 *    false → full clone → caller counts it.
 *    null  → a non-ENOENT error on a resolved repo → caller POISONS (unknown).
 *  Matches Python `_is_shallow`: once `_repo_root` proves a usable work tree,
 *  any later git failure on it is a genuine error, never a benign skip. */
export async function isShallow(fs: GitFs, root: string): Promise<boolean | null> {
  const marker = joinPath(root, ".git", "shallow");
  try {
    await fs.promises.stat(marker);
    return true; // marker present → shallow
  } catch (err) {
    if (isENOENT(err)) return false; // marker absent → full history
    return null; // resolved repo errored on the probe → poison
  }
}

/** Tri-state HEAD probe of a resolved root:
 *    "born"   → HEAD resolves to a commit → proceed to count.
 *    "unborn" → freshly-init'd repo, no commits (real 0). isomorphic-git throws
 *               NotFoundError when HEAD's branch ref does not exist yet.
 *    "error"  → any OTHER failure on a resolved repo → poison (unknown). A git
 *               error must not masquerade as "the operator made no commits."
 *  Matches Python `_head_state` (exit 1 = unborn, other = unknown). */
export async function headState(
  git: GitApi,
  fs: GitFs,
  dir: string,
): Promise<"born" | "unborn" | "error"> {
  try {
    await git.resolveRef({ fs, dir, ref: "HEAD" });
    return "born";
  } catch (err) {
    // instanceof can fail across module realms (lazy import), so also match the
    // error code isomorphic-git stamps on its typed errors.
    if (err instanceof git.Errors.NotFoundError) return "unborn";
    if ((err as { code?: string } | null)?.code === "NotFoundError") return "unborn";
    return "error";
  }
}

/** Distinct SHAs in `root` authored by any email in `emails` within the
 *  half-open window [sinceMs, untilMs). null on a resolved-repo git error
 *  (poison). Faithful port of Python `_repo_shas`. */
async function repoShas(
  git: GitApi,
  fs: GitFs,
  root: string,
  emails: Set<string>,
  sinceMs: number,
  untilMs: number,
): Promise<Set<string> | null> {
  const head = await headState(git, fs, root);
  if (head === "error") return null; // git error after resolving root → unknown
  if (head === "unborn") return new Set(); // unborn HEAD: inspected, zero commits

  let commits: ReadCommitResult[];
  try {
    commits = await git.log({ fs, dir: root, ref: "HEAD" });
  } catch {
    return null; // born HEAD but the log walk errored → poison
  }

  const shas = new Set<string>();
  for (const entry of commits) {
    const author = entry.commit?.author;
    if (author == null) continue;
    if (!emails.has(author.email)) continue;
    const instantMs = author.timestamp * 1000; // epoch seconds → ms, absolute UTC
    if (!Number.isFinite(instantMs)) continue;
    if (sinceMs <= instantMs && instantMs < untilMs) shas.add(entry.oid);
  }
  return shas;
}

// --- public API ------------------------------------------------------------

export interface CountCommitsOptions {
  /** isomorphic-git-compatible fs: the FSA adapter (browser) or node:fs (test). */
  fs: GitFs;
  /** Candidate working directories from the granted repos (may be repo roots or
   *  subdirs — each is resolved with findRoot). Non-git / missing dirs skipped. */
  repoDirs: Iterable<string>;
  /** The operator's selected author emails. Empty → null (no identity). */
  authorEmails: Iterable<string>;
  /** Reporting window, half-open [since, until). */
  since: Date;
  until: Date;
  /** Optional progress hook, called once per resolved repo after it is counted. */
  onProgress?: (done: number, total: number) => void;
  /** Optional cancellation. An abort yields null (unknown), never a partial. */
  signal?: AbortSignal;
}

/** The VES numerator: distinct own-authored commits in [since, until) across
 *  the granted repos, or null when the count cannot be trusted. Faithful port
 *  of Python `count_commits`. */
export async function countCommits(options: CountCommitsOptions): Promise<number | null> {
  const emails = new Set([...options.authorEmails].filter((e) => e));
  if (emails.size === 0) return null; // no git identity → unknown

  const { fs, since, until, signal, onProgress } = options;
  const sinceMs = since.getTime();
  const untilMs = until.getTime();
  if (!Number.isFinite(sinceMs) || !Number.isFinite(untilMs)) return null;

  const git = await loadGit();

  // 1) Resolve granted dirs → repo roots. Skip non-git / shallow; poison only
  //    on a resolved-repo probe error. Dedup roots (many sessions, one repo).
  const roots = new Set<string>();
  for (const dir of options.repoDirs) {
    if (signal?.aborted) return null; // cancelled → unknown, not a partial
    if (!dir) continue; // missing cwd → unattributable, skip
    const root = await findRepoRoot(git, fs, dir);
    if (root == null) continue; // not a git work tree → skip
    const shallow = await isShallow(fs, root);
    if (shallow === null) return null; // resolved repo errored on probe → poison
    if (shallow) continue; // truncated history → cannot trust this repo, skip
    roots.add(root);
  }
  if (roots.size === 0) return null; // nothing verifiable → unknown, not zero

  // 2) Count distinct SHAs across roots; poison on any resolved-repo error.
  const allShas = new Set<string>();
  let done = 0;
  for (const root of roots) {
    if (signal?.aborted) return null;
    const shas = await repoShas(git, fs, root, emails, sinceMs, untilMs);
    if (shas === null) return null;
    for (const sha of shas) allShas.add(sha);
    onProgress?.(++done, roots.size);
  }
  return allShas.size;
}

export interface DiscoverAuthorsResult {
  /** Distinct `user.email` configured across the granted repos (preselect). */
  configured: string[];
  /** Distinct author emails seen in recent HEAD commits, most-frequent first
   *  (the chip choices). Bounded by `recentLimit` commits PER repo. */
  recent: string[];
}

/** Surface the candidate author emails for the chip picker: the locally
 *  configured `user.email`(s) to preselect, plus distinct recent HEAD authors
 *  to offer. Read-only; nothing here is uploaded. Resolution/probe failures are
 *  skipped silently (best-effort discovery, never poisons the UI). */
export async function discoverAuthors(options: {
  fs: GitFs;
  repoDirs: Iterable<string>;
  recentLimit?: number;
}): Promise<DiscoverAuthorsResult> {
  const limit = options.recentLimit ?? 200;
  const git = await loadGit();
  const { fs } = options;

  const roots = new Set<string>();
  for (const dir of options.repoDirs) {
    if (!dir) continue;
    const root = await findRepoRoot(git, fs, dir);
    if (root != null) roots.add(root);
  }

  const configured = new Set<string>();
  const freq = new Map<string, number>();
  for (const root of roots) {
    try {
      const email = await git.getConfig({ fs, dir: root, path: "user.email" });
      if (email) configured.add(email.trim());
    } catch {
      /* no local identity in this repo — skip */
    }
    if ((await headState(git, fs, root)) !== "born") continue;
    try {
      const commits = await git.log({ fs, dir: root, ref: "HEAD" });
      for (const entry of commits.slice(0, limit)) {
        const email = entry.commit?.author?.email;
        if (email) freq.set(email, (freq.get(email) ?? 0) + 1);
      }
    } catch {
      /* unreadable history — skip this repo's authors */
    }
  }

  const recent = [...freq.entries()].sort((a, b) => b[1] - a[1]).map(([email]) => email);
  return { configured: [...configured], recent };
}

// --- browser FSA → fs read-only adapter ------------------------------------
// Browser-only (uses File System Access handles). Not exercised by the Node
// unit tests, which pass node:fs directly; validated by the local Chrome flow
// and e2e. Maps the path strings isomorphic-git builds (`${rootPath}/.git/...`)
// onto a granted FileSystemDirectoryHandle by walking child handles. Read-only:
// write ops throw, so a bug can never mutate the user's repo.

function enoent(path: string): Error & { code: string } {
  const err = new Error(`ENOENT: no such file or directory, '${path}'`) as Error & { code: string };
  err.code = "ENOENT";
  return err;
}

function readonlyError(): never {
  const err = new Error("EROFS: gitcount fs adapter is read-only") as Error & { code: string };
  err.code = "EROFS";
  throw err;
}

function makeStat(kind: "file" | "dir", size: number): GitStats {
  const isDir = kind === "dir";
  return {
    type: kind,
    mode: isDir ? 0o040000 : 0o100644,
    size: isDir ? 0 : size,
    ino: 0,
    dev: 0,
    uid: 0,
    gid: 0,
    ctimeMs: 0,
    mtimeMs: 0,
    ctimeSeconds: 0,
    ctimeNanoseconds: 0,
    mtimeSeconds: 0,
    mtimeNanoseconds: 0,
    isFile: () => !isDir,
    isDirectory: () => isDir,
    isSymbolicLink: () => false,
  };
}

/** Build a read-only isomorphic-git fs over a granted directory handle.
 *  `rootPath` is the path string that corresponds to `root` (the same prefix
 *  the caller uses when building repoDirs, e.g. "/grant"). Paths outside
 *  `rootPath` (a walk above the grant) resolve to ENOENT — which makes
 *  findRoot terminate and the dir get skipped, the correct behavior. */
export function createFsaFs(root: FileSystemDirectoryHandle, rootPath: string): GitFs {
  const base = rootPath.replace(/\/+$/, "");

  // Segments of `path` relative to base, or null if path escapes the grant.
  function relSegments(path: string): string[] | null {
    const norm = path.replace(/\/+$/, "");
    if (norm === base) return [];
    if (!norm.startsWith(base + "/")) return null; // above/outside the grant
    return norm
      .slice(base.length + 1)
      .split("/")
      .filter((s) => s.length > 0 && s !== ".");
  }

  function mapFsaError(err: unknown, path: string): Error {
    // FSA throws DOMException NotFoundError/TypeMismatchError; isomorphic-git
    // keys off err.code === 'ENOENT'. Translate the absence cases.
    const name = (err as { name?: string } | null)?.name;
    if (name === "NotFoundError" || name === "TypeMismatchError") return enoent(path);
    return err as Error;
  }

  async function resolveDir(segments: string[], path: string): Promise<FileSystemDirectoryHandle> {
    let handle = root;
    for (const seg of segments) {
      try {
        handle = await handle.getDirectoryHandle(seg);
      } catch (err) {
        throw mapFsaError(err, path);
      }
    }
    return handle;
  }

  const promises: GitFsPromises = {
    async readFile(path, optionsArg) {
      const segments = relSegments(path);
      if (segments == null || segments.length === 0) throw enoent(path);
      const name = segments[segments.length - 1];
      const dir = await resolveDir(segments.slice(0, -1), path);
      let file: File;
      try {
        const fileHandle = await dir.getFileHandle(name);
        file = await fileHandle.getFile();
      } catch (err) {
        throw mapFsaError(err, path);
      }
      const bytes = new Uint8Array(await file.arrayBuffer());
      const encoding = typeof optionsArg === "string" ? optionsArg : optionsArg?.encoding;
      if (encoding === "utf8" || encoding === "utf-8") return new TextDecoder().decode(bytes);
      return bytes;
    },

    async stat(path) {
      const segments = relSegments(path);
      if (segments == null) throw enoent(path);
      if (segments.length === 0) return makeStat("dir", 0); // the grant root itself
      const name = segments[segments.length - 1];
      const parent = await resolveDir(segments.slice(0, -1), path);
      // Try file first, then directory; neither → ENOENT.
      try {
        const fileHandle = await parent.getFileHandle(name);
        const file = await fileHandle.getFile();
        return makeStat("file", file.size);
      } catch {
        /* not a file — fall through to directory probe */
      }
      try {
        await parent.getDirectoryHandle(name);
        return makeStat("dir", 0);
      } catch (err) {
        throw mapFsaError(err, path);
      }
    },

    async lstat(path) {
      return promises.stat(path); // FSA exposes no symlinks
    },

    async readdir(path) {
      const segments = relSegments(path);
      if (segments == null) throw enoent(path);
      const dir = await resolveDir(segments, path);
      const names: string[] = [];
      // FileSystemDirectoryHandle is async-iterable over its entry keys.
      for await (const name of (dir as unknown as { keys(): AsyncIterable<string> }).keys()) {
        names.push(name);
      }
      return names;
    },

    async readlink() {
      const err = new Error("EINVAL: not a symlink") as Error & { code: string };
      err.code = "EINVAL";
      throw err;
    },

    // Write ops: present so isomorphic-git's bindFs can bind them, but the
    // adapter is strictly read-only — invoking any of these throws.
    writeFile: () => readonlyError(),
    mkdir: () => readonlyError(),
    unlink: () => readonlyError(),
    symlink: () => readonlyError(),
    rmdir: () => readonlyError(),
  };

  return { promises };
}
